HTTP Probe: Removes Accept-Encoding header from http probe

fonsecas72 opened the pull request:
/sig node
/sig network
/kind cleanup

**What this PR does / why we need it**:

The `Accept-Encoding` header is added by default because compression is enabled by default. 
By removing it we are reducing the sent payload.

Also it relates with https://github.com/kubernetes/kubernetes/pull/95641

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Changed: default "Accept-Encoding" header removed from HTTP probes. See https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#http-probes

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

fonsecas72 commented:
/assign @SergeyKanzhelev @dims 

Please help me with the ok-to-test

SergeyKanzhelev commented:
/ok-to-test

fonsecas72 commented:
Thank you
/retest

SergeyKanzhelev commented on pkg/probe/http/http.go:
```diff
@@ -70,6 +70,7 @@ type httpProber struct {
 
 // Probe returns a ProbeRunner capable of running an HTTP check.
 func (pr httpProber) Probe(url *url.URL, headers http.Header, timeout time.Duration) (probe.Result, string, error) {
+	pr.transport.DisableCompression = true // removes Accept-Encoding header
```
interesting... So you disable the header using `DisableCompression`, but can add header by simply adding it to the headers collection. After a brief internet search I wonder if Accept-Encoding needs to be set to `identity` as well (https://groups.google.com/g/golang-nuts/c/ms8QNCzew8Q)? Cannot find any call to dumping http request anywhere, maybe you saw it somewhere?

fonsecas72 replied:
`If no Accept-Encoding field is in the request, any content-coding is considered acceptable by the user agent.`

We should not need to set "Accept-Encoding: identity", in fact my understanding is that having no Accept-Encoding is different from "Accept-Encoding: identify", as with the former should be equivalent to "Accept-Encoding: *" 
~> "give-me no-compression" vs "give-me anything" (?)

( https://tools.ietf.org/html/rfc7231#section-5.3.4 )

Also to me, it seems we may have 2 different questions:
* do we just want to reduce the payload ?
* do we want to disable the compression by default ?

SergeyKanzhelev replied:
got it. Thank you. I misunderstood the article than,

With your explanation I think removing the headers benefits both - less payload on request and response can still be gzip for apps worried about network and non-gzip for ones that are CPU bound.

fonsecas72 replied:
yeah, i think so. Seems server should fallback to gzip if no "identify" is available, so basically means that it should not break in case the server only knows to respond with gzip encoding 🤔 

> Note: If the request does not include an Accept-Encoding field,
>       and if the "identity" content-coding is unavailable, then
>       content-codings commonly understood by HTTP/1.0 clients (i.e.,
>       "gzip" and "compress") are preferred; some older clients
>       improperly display messages sent with other content-codings.  The
>       server might also make this decision based on information about
>       the particular user-agent or client

https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html

sftim replied:
*aside* If you want to go wild, send `TE: gzip` and allow HTTP/1.1 gzip _transfer encoding_.
(Not many origins support that). More details in https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/TE

SergeyKanzhelev commented on pkg/probe/http/http_test.go:
```diff
@@ -158,13 +158,21 @@ func TestHTTPProbeChecker(t *testing.T) {
+		{
+			handler: headerEchoHandler,
+			reqHeaders: http.Header{
+				"Accept-Encoding": {"gzip"},
```
what will happen with the empty string for the `Accept-Encoding`? I wonder for the behavor before this change and after. Would any users already setting it to empty experience a different behavior?

fonsecas72 replied:
https://tools.ietf.org/html/rfc7231#section-5.3.4

`
If the representation has no content-coding, then it is  acceptable by default unless specifically excluded by the Accept-Encoding field stating either "identity;q=0" or "*;q=0" without a more specific entry for "identity"
`

My understanding is that the server should translate an "Accept-Encoding: " into "Accept-Encoding: identity"

So I don't anticipate any breaking change in this situation

SergeyKanzhelev replied:
I was mostly asking from the client perspective. If empty string was set for `"Accept-Encoding"` in yaml, will it translate to no header or to empty string header before and after this PR?

fonsecas72 replied:
ah, it should translate to empty string as empty string is a valid "Accept-Encoding". I will add a test

fonsecas72 replied:
done 👍  ty

SergeyKanzhelev commented on pkg/probe/http/http_test.go:
```diff
@@ -158,13 +158,21 @@ func TestHTTPProbeChecker(t *testing.T) {
+		{
+			handler: headerEchoHandler,
+			reqHeaders: http.Header{
+				"Accept-Encoding": {"gzip"},
```
also please add a test for the non-strandard encoding value

fonsecas72 replied:
👍 done, ty

fonsecas72 commented:
/retest

fonsecas72 committed:
HTTP Prove: Removes Accept-Encoding header from http probe

The Accept-Encoding header is added by default because compression is enabled by default. By removing it we are reducing the sent payload.

SergeyKanzhelev commented:
@fonsecas72 please update the release note - the link to PR is not correct there. Perhaps simply remove it or start a new PR to the website.

/lgtm

/assign @thockin 

Tim, asking you here since you commented on https://github.com/kubernetes/kubernetes/issues/95596 . Do you see any issues removing this header from the list of default headers? Missing header is equivalent of "whatever, prefer gzip" (see this comment: https://github.com/kubernetes/kubernetes/pull/96127#discussion_r520141542). And it minimizes the default http requests payload.

thockin commented:
This seems reasonable to me - you read the specs and if you're happy, I am.

thockin commented:
/approve
